source/simulator/stokes_matrix_free.cc

```diff
@@ -286,11 +293,29 @@ namespace aspect
       if (do_solve_A == true)
         {
           SolverControl solver_control(1000, utmp.block(0).l2_norm() * A_block_tolerance);
-          SolverCG<dealii::LinearAlgebra::distributed::Vector<double>> solver(solver_control);
+          PrimitiveVectorMemory<dealii::LinearAlgebra::distributed::Vector<double>> mem;
+
           try
             {
-              solver.solve(A_block, dst.block(0), utmp.block(0),
-                           A_block_preconditioner);
+              if (A_block_is_symmetric)
+                {
+                  SolverCG<dealii::LinearAlgebra::distributed::Vector<double>> solver(solver_control,mem);
+                  solver.solve(A_block, dst.block(0), utmp.block(0),
+                               A_block_preconditioner);
+                }
+              else
+                {
+                  // Use BiCGStab for non-symmetric matrices.
+                  // Do not compute the exact residual, as this
+                  // is more expensive, and we only need an approximate solution.
+                  SolverBicgstab<dealii::LinearAlgebra::distributed::Vector<double>>
+                  solver(solver_control,
+                         mem,
+                         SolverBicgstab<dealii::LinearAlgebra::distributed::Vector<double>>::AdditionalData(/*exact_residual=*/ false));
+                  solver.solve(A_block, dst.block(0), utmp.block(0),
+                               A_block_preconditioner);
+                }
```

Comment:
This is a question for @tjhei: What exactly is the matrix-free preconditioner we are using in the situations where A is not symmetric? Does the matrix-free preconditioner even know about the stabilization term? And is the preconditioner able to deal with nonsymmetric matrices?

Reply:
Yes, we implemented free surface terms. We are using Chebyshev, which should work okay for non-symmetric operators
https://github.com/geodynamics/aspect/blob/512911f72be1060a85974f1886dcb9ac5909435c/source/simulator/stokes_matrix_free.cc#L1722